pgrx-pg-sys/src/submodules/thread_check.rs

```diff
@@ -38,8 +40,60 @@ pub(crate) fn check_active_thread() {
     }
 }
 
+/// Use OS-specific mechanisms to detect if we're the process main thread, if
+/// supported on the OS. Should return `None` when unsupported, or if there's an
+/// error.
+///
+/// Concretely, it is very important that this not return `Some(false)`
+/// incorrectly, but the other values are less important. Callers generally
+/// should compare the result against `Some(false)`.
+fn is_os_main_thread() -> Option<bool> {
+    #[cfg(any(target_os = "macos", target_os = "openbsd", target_os = "freebsd"))]
+    return unsafe {
+        match libc::pthread_main_np() {
+            1 => Some(true),
+            0 => Some(false),
+            // Note that this returns `-1` in some error conditions.
+            //
+            // In these cases we are almost certainly not the main thread, but
+            // we don't know -- it's better for this function to return `None`
+            // in cases of uncertainty.
```

Comment:
They claim this is about the thread not finishing initialization? https://opensource.apple.com/source/libpthread/libpthread-218.1.3/man/pthread_main_np.3.auto.html